src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/Razor/ExtractToNewComponentCodeActionProvider.cs

```diff
@@ -47,51 +51,169 @@ public Task<ImmutableArray<RazorVSInternalCodeAction>> ProvideAsync(RazorCodeAct
+        {
+            endOwner = previousSibling;
+        }
+
+        return endOwner.FirstAncestorOrSelf<MarkupElementSyntax>();
+    }
+
+    private static ExtractToNewComponentCodeActionParams CreateInitialActionParams(RazorCodeActionContext context, MarkupElementSyntax startElementNode, string @namespace)
+    {
+        return new ExtractToNewComponentCodeActionParams
         {
             Uri = context.Request.TextDocument.Uri,
-            ExtractStart = componentNode.Span.Start,
-            ExtractEnd = componentNode.Span.End,
+            ExtractStart = startElementNode.Span.Start,
+            ExtractEnd = startElementNode.Span.End,
             Namespace = @namespace
         };
+    }
 
-        var resolutionParams = new RazorCodeActionResolutionParams()
+    /// <summary>
+    /// Processes a multi-point selection to determine the correct range for extraction.
+    /// </summary>
+    /// <param name="startElementNode">The starting element of the selection.</param>
+    /// <param name="endElementNode">The ending element of the selection, if it exists.</param>
+    /// <param name="actionParams">The parameters for the extraction action, which will be updated.</param>
+    private static void ProcessMultiPointSelection(MarkupElementSyntax startElementNode, MarkupElementSyntax? endElementNode, ExtractToNewComponentCodeActionParams actionParams)
+    {
+        // If there's no end element, we can't process a multi-point selection
+        if (endElementNode is null)
         {
-            Action = LanguageServerConstants.CodeActions.ExtractToNewComponentAction,
-            Language = LanguageServerConstants.CodeActions.Languages.Razor,
-            Data = actionParams,
-        };
+            return;
+        }
 
-        var codeAction = RazorCodeActionFactory.CreateExtractToNewComponent(resolutionParams);
+        var startNodeContainsEndNode = endElementNode.Ancestors().Any(node => node == startElementNode);
 
-        return Task.FromResult<ImmutableArray<RazorVSInternalCodeAction>>([codeAction]);
+        // If the start element is an ancestor, keep the original end; otherwise, use the end of the end element
+        if (!startNodeContainsEndNode)
+        {
+            actionParams.ExtractEnd = endElementNode.Span.End;
+        }
+
+        // If the start element is not an ancestor of the end element, we need to find a common parent
+        // This conditional handles cases where the user's selection spans across different levels of the DOM.
+        // For example:
+        //   <div>
+        //     <span>
+        //      Selected text starts here<p>Some text</p>
+        //     </span>
+        //     <span>
+        //       <p>More text</p>
+        //     </span>
+        //     Selected text ends here <span></span>
+        //   </div>
+        // In this case, we need to find the smallest set of complete elements that covers the entire selection.
```

Comment:
I love this comment <3

BUT... it would be clearer to just write things as we would for test input, with the same markers, and also indicate what the expected resulting selection would be. We all need to be used to reading that syntax anyway.

eg (and correct me if my guess is wrong)
```
//   <div>
//     {|result:<span>
//      {|selection:<p>Some text</p>
//     </span>
//     <span>
//       <p>More text</p>
//     </span>
//     <span></span>|}|}
//   </div>
```